File: backend/app/core/crypto.py
"""libsodium primitives shared by the token-decrypt path (app/services/token_service.py)
and the local admin-seed bootstrap script. Everyday user vault crypto (master password
KDF, private-key wrapping, sharing) happens client-side in the browser - this module is
NOT used on the normal request path for that, only for the two exceptions above."""

import nacl.bindings
import nacl.encoding
import nacl.hash
import nacl.pwhash
import nacl.public
import nacl.secret
import nacl.utils

AEAD_KEY_BYTES = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_KEYBYTES
AEAD_NONCE_BYTES = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES

# Argon2id defaults for the seed-admin bootstrap script only (see pwhash_argon2id).
# The browser uses its own, independently-configured ops/mem limits for real users -
# these constants never gate normal login/unlock.
PWHASH_SALT_BYTES = nacl.pwhash.argon2id.SALTBYTES
PWHASH_OPSLIMIT_DEFAULT = nacl.pwhash.argon2id.OPSLIMIT_MODERATE
PWHASH_MEMLIMIT_DEFAULT = nacl.pwhash.argon2id.MEMLIMIT_MODERATE


def generate_keypair() -> tuple[bytes, bytes]:
    """Returns (private_key, public_key), 32 bytes each (X25519)."""
    key = nacl.public.PrivateKey.generate()
    return bytes(key), bytes(key.public_key)


def pwhash_argon2id(password: bytes, salt: bytes, ops_limit: int, mem_limit: int) -> bytes:
    """Derives a 32-byte key from a password. `salt` must be nacl.pwhash.argon2id.SALTBYTES long."""
    return nacl.pwhash.argon2id.kdf(
        AEAD_KEY_BYTES, password, salt, opslimit=ops_limit, memlimit=mem_limit
    )


def secretbox_encrypt(plaintext: bytes, key: bytes) -> tuple[bytes, bytes]:
    """crypto_secretbox (XSalsa20-Poly1305) - used specifically for wrapping a user's
    private key with their master-password-derived stretch key, per the crypto design.
    Distinct from aead_encrypt below, which is XChaCha20-Poly1305-IETF and used for
    everything else (resource/note/token content)."""
    nonce = nacl.utils.random(nacl.secret.SecretBox.NONCE_SIZE)
    box = nacl.secret.SecretBox(key)
    ciphertext = box.encrypt(plaintext, nonce).ciphertext
    return ciphertext, nonce


def secretbox_decrypt(ciphertext: bytes, nonce: bytes, key: bytes) -> bytes:
    box = nacl.secret.SecretBox(key)
    return box.decrypt(ciphertext, nonce)


def aead_encrypt(plaintext: bytes, key: bytes) -> tuple[bytes, bytes]:
    """XChaCha20-Poly1305-IETF AEAD encrypt. Returns (ciphertext, nonce)."""
    nonce = nacl.utils.random(AEAD_NONCE_BYTES)
    ciphertext = nacl.bindings.crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, None, nonce, key)
    return ciphertext, nonce


def aead_decrypt(ciphertext: bytes, nonce: bytes, key: bytes) -> bytes:
    """Raises nacl.exceptions.CryptoError on tampered/wrong-key ciphertext."""
    return nacl.bindings.crypto_aead_xchacha20poly1305_ietf_decrypt(ciphertext, None, nonce, key)


def keyed_blake2b(message: bytes, key: bytes, digest_size: int = 32) -> bytes:
    return nacl.hash.blake2b(message, digest_size=digest_size, key=key, encoder=nacl.encoding.RawEncoder)


TOKEN_WRAP_KEY_DOMAIN = b"infrawarden-token-wrap-v1:"


def derive_token_wrap_key(token_id: str, token_secret: bytes) -> bytes:
    """Derives the key that wraps a client's data key for a scoped API token.
    token_id (the token's own UUID, as its string form) is folded in as domain
    separation context so the same token_secret bytes can never accidentally
    derive the same key for two different tokens. Must match the browser's
    lib/crypto.ts derivation exactly - see docs/ARCHITECTURE.md."""
    return keyed_blake2b(TOKEN_WRAP_KEY_DOMAIN + token_id.encode(), key=token_secret)


def seal_for_public_key(plaintext: bytes, public_key: bytes) -> bytes:
    """crypto_box_seal - wraps `plaintext` (e.g. a client data key) so only the holder
    of the matching private key can open it. Self-contained ciphertext, no nonce needed."""
    box = nacl.public.SealedBox(nacl.public.PublicKey(public_key))
    return box.encrypt(plaintext)


def unseal_with_private_key(sealed: bytes, private_key: bytes) -> bytes:
    box = nacl.public.SealedBox(nacl.public.PrivateKey(private_key))
    return box.decrypt(sealed)

File: frontend/src/lib/crypto.ts
// All client-side vault cryptography lives here. Mirrors backend/app/core/crypto.py -
// same libsodium primitives on both sides so wire formats match exactly. See
// docs/ARCHITECTURE.md for why each primitive was chosen.
import sodium from "libsodium-wrappers-sumo";

let readyPromise: Promise<typeof sodium> | null = null;

async function ready(): Promise<typeof sodium> {
  if (!readyPromise) {
    readyPromise = sodium.ready.then(() => sodium);
  }
  return readyPromise;
}

export interface KdfParams {
  opsLimit: number;
  memLimit: number;
}

// Matches backend PWHASH_OPSLIMIT_DEFAULT / PWHASH_MEMLIMIT_DEFAULT (argon2id MODERATE).
export const DEFAULT_KDF_PARAMS: KdfParams = {
  opsLimit: 3,
  memLimit: 268435456,
};

export interface Keypair {
  publicKey: Uint8Array;
  privateKey: Uint8Array;
}

export async function generateKeypair(): Promise<Keypair> {
  const s = await ready();
  const kp = s.crypto_box_keypair();
  return { publicKey: kp.publicKey, privateKey: kp.privateKey };
}

/** A fresh random symmetric data key for a new client vault. */
export async function generateDataKey(): Promise<Uint8Array> {
  const s = await ready();
  return s.randombytes_buf(32);
}

export async function generateKdfSalt(): Promise<Uint8Array> {
  const s = await ready();
  return s.randombytes_buf(s.crypto_pwhash_SALTBYTES);
}

/** Derives the 32-byte stretch key from the master password. Never leaves the caller. */
export async function deriveStretchKey(
  password: string,
  salt: Uint8Array,
  params: KdfParams = DEFAULT_KDF_PARAMS,
): Promise<Uint8Array> {
  const s = await ready();
  return s.crypto_pwhash(
    32,
    password,
    salt,
    params.opsLimit,
    params.memLimit,
    s.crypto_pwhash_ALG_ARGON2ID13,
  );
}

/** The login verifier sent to the server - never the password or stretch key itself. */
export async function deriveAuthHash(password: string, stretchKey: Uint8Array): Promise<string> {
  const s = await ready();
  const digest = s.crypto_generichash(32, s.from_string(password), stretchKey);
  return s.to_hex(digest);
}

/** crypto_secretbox (XSalsa20-Poly1305) - wraps the private key with the stretch key. */
export async function wrapPrivateKey(
  privateKey: Uint8Array,
  stretchKey: Uint8Array,
): Promise<{ ciphertext: Uint8Array; nonce: Uint8Array }> {
  const s = await ready();
  const nonce = s.randombytes_buf(s.crypto_secretbox_NONCEBYTES);
  const ciphertext = s.crypto_secretbox_easy(privateKey, nonce, stretchKey);
  return { ciphertext, nonce };
}

export async function unwrapPrivateKey(
  ciphertext: Uint8Array,
  nonce: Uint8Array,
  stretchKey: Uint8Array,
): Promise<Uint8Array> {
  const s = await ready();
  return s.crypto_secretbox_open_easy(ciphertext, nonce, stretchKey);
}

/** crypto_box_seal - wraps a symmetric key (e.g. a client data key) to a public key. */
export async function sealForPublicKey(plaintext: Uint8Array, publicKey: Uint8Array): Promise<Uint8Array> {
  const s = await ready();
  return s.crypto_box_seal(plaintext, publicKey);
}

export async function unsealWithKeypair(
  sealed: Uint8Array,
  publicKey: Uint8Array,
  privateKey: Uint8Array,
): Promise<Uint8Array> {
  const s = await ready();
  return s.crypto_box_seal_open(sealed, publicKey, privateKey);
}

/** XChaCha20-Poly1305-IETF AEAD - used for resource/note content and token-wrapped data keys. */
export async function aeadEncrypt(
  plaintext: Uint8Array,
  key: Uint8Array,
): Promise<{ ciphertext: Uint8Array; nonce: Uint8Array }> {
  const s = await ready();
  const nonce = s.randombytes_buf(s.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES);
  const ciphertext = s.crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, null, null, nonce, key);
  return { ciphertext, nonce };
}

export async function aeadDecrypt(ciphertext: Uint8Array, nonce: Uint8Array, key: Uint8Array): Promise<Uint8Array> {
  const s = await ready();
  return s.crypto_aead_xchacha20poly1305_ietf_decrypt(null, ciphertext, null, nonce, key);
}

/** Convenience wrappers for resource/note content, which is always JSON before encryption. */
export async function encryptJson(
  value: unknown,
  key: Uint8Array,
): Promise<{ ciphertext: Uint8Array; nonce: Uint8Array }> {
  const s = await ready();
  return aeadEncrypt(s.from_string(JSON.stringify(value)), key);
}

export async function decryptJson<T>(ciphertext: Uint8Array, nonce: Uint8Array, key: Uint8Array): Promise<T> {
  const s = await ready();
  const plaintext = await aeadDecrypt(ciphertext, nonce, key);
  return JSON.parse(s.to_string(plaintext)) as T;
}

export async function toBase64(bytes: Uint8Array): Promise<string> {
  const s = await ready();
  return s.to_base64(bytes, s.base64_variants.ORIGINAL);
}

export async function fromBase64(b64: string): Promise<Uint8Array> {
  const s = await ready();
  return s.from_base64(b64, s.base64_variants.ORIGINAL);
}
